examples/mnist.py

```diff
@@ -258,80 +237,73 @@ def main(model='mlp', num_epochs=500):
     test_prediction = lasagne.layers.get_output(network, deterministic=True)
     test_loss = lasagne.objectives.categorical_crossentropy(test_prediction,
                                                             target_var)
-    test_loss = test_loss.mean()
     # As a bonus, also create an expression for the classification accuracy:
-    test_acc = T.mean(T.eq(T.argmax(test_prediction, axis=1), target_var),
-                      dtype=theano.config.floatX)
+    test_err = T.sum(T.neq(T.argmax(test_prediction, axis=1), target_var),
+                     dtype=theano.config.floatX)
 
     # Compile a function performing a training step on a mini-batch (by giving
     # the updates dictionary) and returning the corresponding training loss:
-    train_fn = theano.function([input_var, target_var], loss, updates=updates)
+    train_fn = theano.function([input_var, target_var], loss.sum(),
+                               updates=updates)
 
     # Compile a second function computing the validation loss and accuracy:
-    val_fn = theano.function([input_var, target_var], [test_loss, test_acc])
+    val_fn = theano.function([input_var, target_var],
+                             [test_loss.sum(), test_err.sum()])
 
     # Finally, launch the training loop.
     print("Starting training...")
     # We iterate over epochs:
     for epoch in range(num_epochs):
         # In each epoch, we do a full pass over the training data:
-        train_err = 0
-        train_batches = 0
+        train_loss = 0
         start_time = time.time()
-        for batch in iterate_minibatches(X_train, y_train, 500, shuffle=True):
-            inputs, targets = batch
-            train_err += train_fn(inputs, targets)
-            train_batches += 1
+        for inputs, targets in train_ds.batch_iterator(
+                500, shuffle_rng=lasagne.random.get_rng()):
+            train_loss += train_fn(inputs, targets)
 
         # And a full pass over the validation data:
+        val_loss = 0
         val_err = 0
-        val_acc = 0
-        val_batches = 0
-        for batch in iterate_minibatches(X_val, y_val, 500, shuffle=False):
-            inputs, targets = batch
-            err, acc = val_fn(inputs, targets)
+        for inputs, targets in val_ds.batch_iterator(500):
+            loss, err = val_fn(inputs, targets)
+            val_loss += loss
             val_err += err
-            val_acc += acc
-            val_batches += 1
 
         # Then we print the results for this epoch:
         print("Epoch {} of {} took {:.3f}s".format(
             epoch + 1, num_epochs, time.time() - start_time))
-        print("  training loss:\t\t{:.6f}".format(train_err / train_batches))
-        print("  validation loss:\t\t{:.6f}".format(val_err / val_batches))
-        print("  validation accuracy:\t\t{:.2f} %".format(
-            val_acc / val_batches * 100))
+        print("  training loss:\t\t{:.6f}".format(
+            train_loss / X_train.shape[0]))
```

Comment:
This is only correct if `len(X_train) % batchsize == 0`. Dividing the sum of batch averages by the batch count is safer.